Wikipedia/Code/NavigationBarHider.swift

```diff
@@ -32,23 +32,43 @@ public class NavigationBarHider: NSObject {
     
     fileprivate var isUserScrolling: Bool = false
     fileprivate var isScrollingToTop: Bool = false
+    fileprivate var isProgramaticallyScrolling: Bool = false
     var initialScrollY: CGFloat = 0
     var initialNavigationBarPercentHidden: CGFloat = 0
+
+    private func setDataForScroll(yOffset: CGFloat, topContentInset: CGFloat) {
+        guard let navigationBar = navigationBar else {
+            return
+        }
+        initialScrollY = yOffset + topContentInset
+        initialNavigationBarPercentHidden = navigationBar.navigationBarPercentHidden
+    }
+
+    public func setIsProgramaticallyScrolling(_ isProgramaticallyScrolling: Bool, yOffset: CGFloat, topContentInset: CGFloat) {
+        self.isProgramaticallyScrolling = isProgramaticallyScrolling
+
+        if isProgramaticallyScrolling {
+            setDataForScroll(yOffset: yOffset, topContentInset: topContentInset)
+        }
+    }
     
     @objc public func scrollViewWillBeginDragging(_ scrollView: UIScrollView) {
-        guard let navigationBar = navigationBar, navigationBar.isInteractiveHidingEnabled else {
+        guard navigationBar?.isInteractiveHidingEnabled == true else {
             return
         }
         isUserScrolling = true
-        initialScrollY = scrollView.contentOffset.y + scrollView.contentInset.top
-        initialNavigationBarPercentHidden = navigationBar.navigationBarPercentHidden
+        setDataForScroll(yOffset: scrollView.contentOffset.y, topContentInset: scrollView.contentInset.top)
     }
 
     @objc public func scrollViewDidScroll(_ scrollView: UIScrollView) {
         guard let navigationBar = navigationBar else {
             return
         }
-        
+
+        /// When programatically scrolling on a just-created scroll view, we need a layout pass to get accurate heights.
+        scrollView.layoutIfNeeded()
```

Comment:
If it only seems needed when programmatically scrolling, might want to wrap these in `if isProgramaticallyScrolling {` to limit unneeded layout passes in other states.

Reply:
Good call, updated!